drivers/aip31068/aip31068.c

```diff
@@ -0,0 +1,516 @@
+ *
+ * @param[in] dev   Device descriptor of the AIP31068
+ * @param[in] data  Data to write
+ * @param[in] len   Length of the data
+ *
+ * @return  AIP31068_OK on success
+ * @return  -AIP31068_ERROR_I2C if acquiring of I2C bus fails
+ * @return  -EIO When slave device doesn't ACK the byte
+ * @return  -ENXIO When no devices respond on the address sent on the bus
+ * @return  -ETIMEDOUT When timeout occurs before device's response
+ * @return  -EINVAL When an invalid argument is given
+ * @return  -EOPNOTSUPP When MCU driver doesn't support the flag operation
+ * @return  -EAGAIN When a lost bus arbitration occurs
+ */
+static int _device_write(aip31068_t *dev, uint8_t *data, uint8_t len);
+
+/**
+ * @brief Initialize controller for progress bar feature.
+ *
+ * @param[in] dev   Device descriptor of the AIP31068
+ * @param[in] row   Row where the progress bar is displayed
+ */
+static void _init_progress_bar(aip31068_t* dev, uint8_t row);
+
+/**
+ * Custom character 1 x 8 bar for progress bar.
+ */
+uint8_t custom_char_progress_bar_1[8] = { 16, 16, 16, 16, 16, 16, 16, 16 };
+
+/**
+ * Custom character 2 x 8 bar for progress bar.
+ */
+uint8_t custom_char_progress_bar_2[8] = { 24, 24, 24, 24, 24, 24, 24, 24 };
+
+/**
+ * Custom character 3 x 8 bar for progress bar.
+ */
+uint8_t custom_char_progress_bar_3[8] = { 28, 28, 28, 28, 28, 28, 28, 28 };
+
+/**
+ * Custom character 4 x 8 bar for progress bar.
+ */
+uint8_t custom_char_progress_bar_4[8] = { 30, 30, 30, 30, 30, 30, 30, 30 };
+
+/**
+ * Custom character 5 x 8 bar for progress bar.
+ */
+uint8_t custom_char_progress_bar_5[8] = { 31, 31, 31, 31, 31, 31, 31, 31 };
+
+int aip31068_init(aip31068_t *dev, const aip31068_params_t *params)
+{
+    assert(dev);
+    assert(params);
+
+    dev->params = *params;
+    dev->_curr_display_control = 0;
+    dev->_curr_entry_mode_set = 0;
+
+    i2c_init(dev->params.i2c_dev);
+
+    uint8_t _function_set = 0;
+
+    /* configure bit mode */
+    if (params->bit_mode == BITMODE_8_BIT) {
+        _function_set |= (1 << BIT_FUNCTION_SET_BITMODE);
+    }
+
+    /* configure line count */
+    if (params->row_count >= 2) {
+        _function_set |= (1 << BIT_FUNCTION_SET_LINECOUNT);
+    }
+
+    /* configure character size */
+    if (params->font_size == FONT_SIZE_5x10) {
+        _function_set |= (1 << BIT_FUNCTION_SET_FONTSIZE);
+    }
+
+    /* begin of initialization sequence (page 20 in the datasheet) */
+    xtimer_usleep(50 * US_PER_MS);
+
+    int rc = 0;
+    int c = 0;
+
+    /* send function set command sequence */
+    do {
+        rc = _command(dev, CMD_FUNCTION_SET | _function_set);
+        xtimer_usleep(5 * US_PER_MS);
+
+        printf("A rc = %d\n", rc);
+        c++;
+
+        if (c > 10) {
+            return AIP31068_ERROR_I2C;
+        }
+    } while (rc != 0); // TODO: get rid of the loop if possible
+
+    /* second try */
+    rc = _command(dev, CMD_FUNCTION_SET | _function_set);
+    if (rc < 0) {
+        return rc;
+    }
```

Comment:
`/* second try */` and `/* third go */` can indeed be removed because the loop already does it. It was a bit confusing for me because the aip31068 has the exact same API like the HD44780 but the initilization seems to differ. For the HD44780 this would be necessary, but for the aip31068 one successful call is enough (therefor the loop is necessary at the moment unfortunately).

Reply:
Ok, then lets remove the extra loops.